File: src/main/scala/tdw/spark/examples/TDBankDStreamTest.scala
package com.tencent.tdw.spark.examples

import com.tencent.tdw.spark.examples.util.GeneralArgParser
import com.tencent.tdw.spark.toolkit.tdbank.{HippoReceiverConfig, TubeReceiverConfig, TDBankProvider}
import org.apache.spark.SparkConf
import org.apache.spark.streaming.{Seconds, StreamingContext}

/**
 * Created by sharkdtu
 * 2015/9/9
 */
object TDBankDStreamTest {
  def main(args: Array[String]) {
    val sparkConf = new SparkConf()
    val streamingContext = new StreamingContext(sparkConf, Seconds(30))

    val cmdArgs = new GeneralArgParser(args)

    val master = cmdArgs.getStringValue("master")
    val group = cmdArgs.getStringValue("group")
    val topic = cmdArgs.getStringValue("topic")
    val tids = cmdArgs.getCommaSplitArrayValue("tids")

    val tdbank = new TDBankProvider(streamingContext)
    // From Tube
    val receiverConf = new TubeReceiverConfig().setMaster(master)
      .setGroup(group).setTids(tids).setTopic(topic)
    // From hippo
    // val receiverConf = new HippoReceiverConfig().setControllerAddrs(addrs)
    //  .setGroup(group).setTopic(topic)
    val stream = tdbank.bytesStream(receiverConf, numReceiver = 1)

//    stream.foreachRDD{ rdd =>
//      val ret = rdd.take(2)
//      for(ba <- ret){
//        println(new String(ba))
//      }
//    }

    val cnt = stream.count().map{ x =>
      "count at " + System.currentTimeMillis + ": " + x
    }
    cnt.print()
    streamingContext.start()
    streamingContext.awaitTermination()
  }
}

File: python/src/tdbank_dstream_test.py
from __future__ import print_function

import sys

from pyspark import SparkContext
from pyspark.streaming import StreamingContext
from pytoolkit import TDBankProvider
from pytoolkit import TubeReceiverConfig

if __name__ == "__main__":
    """
    Usage: tdbank_dstream_test.py <master> <group> <topic> <tid>
    """
    if len(sys.argv) < 5:
        print("Usage: tdw_rdd_test.py <master> <group> <topic> <tid>")
        exit(1)

    master = sys.argv[1].strip()
    group = sys.argv[2].strip()
    topic = sys.argv[3].strip()
    tids = sys.argv[4].strip().split(",")

    sc = SparkContext(appName="tdbank-dstream-test")
    ssc = StreamingContext(sc, 1)
    tdbank = TDBankProvider(ssc)
    receiver_conf = TubeReceiverConfig()\
        .setMaster(master)\
        .setGroup(group)\
        .setTopic(topic)\
        .setTids(tids)
    stream = tdbank.textStream(receiver_conf, 1)

    def process(time, rdd):
        print("========= %s =========" % str(time))
        strs = rdd.take(1)
        print(strs)
        print("=====================")

    stream.foreachRDD(process)
    ssc.start()
    ssc.awaitTermination()
